controller/api/destination/endpoint_translator.go

```diff
@@ -79,33 +99,122 @@ func newEndpointTranslator(
 		availableEndpoints,
 		filteredSnapshot,
 		stream,
+		endStream,
 		log,
+		make(chan interface{}, updateQueueCapacity),
+		make(chan struct{}),
 		sync.Mutex{},
 	}
 }
 
 func (et *endpointTranslator) Add(set watcher.AddressSet) {
+	et.enqueueUpdate(&addUpdate{set})
+}
+
+func (et *endpointTranslator) Remove(set watcher.AddressSet) {
+	et.enqueueUpdate(&removeUpdate{set})
+}
+
+func (et *endpointTranslator) NoEndpoints(exists bool) {
+	et.enqueueUpdate(&noEndpointsUpdate{exists})
+}
+
+// Add, Remove, and NoEndpoints are called from a client-go informer callback
+// and therefore must not block. For each of these, we enqueue an update in
+// a channel so that it can be processed asyncronously. To ensure that enqueuing
+// does not block, we first check to see if there is capacity in the buffered
+// channel. If there is not, we drop the update and signal to the stream that
+// it has fallen too far behind and should be closed.
+// This function is protected by a mutex to ensure that the queue does not
+// become full after we detect that it has capacity. Since this guarantees that
+// sending to the channel will not block, the amount of time the mutex will be
+// held is small and will not substantially block the informer callback.
+func (et *endpointTranslator) enqueueUpdate(update interface{}) {
 	et.mu.Lock()
 	defer et.mu.Unlock()
+	if len(et.updates) == cap(et.updates) {
+		// Ensure that we send the end stream signal only once. If the signal
+		// has alread been sent, the stream is closing and we do not need to
+		// signal again.
+		if et.endStream != nil {
+			et.log.Error("endpoint update queue full; ending stream")
+			et.endStream <- struct{}{}
+			et.endStream = nil
+		}
+		return
+	}
+	et.updates <- update
+}
```

Comment:
How about using a select instead to avoid the mutex lock (would require changing the type of `endStream` to be bidirectional?
```go
select {
case et.updates <- update:
default:
    select {
    case <-endStream:
    default:
      et.log.Error("endpoint update queue full; ending stream")
      close(endStream)
    }
}
```

Reply:
great suggestion, thanks!